### PR TITLE
[2783] printf command

### PR DESCRIPTION
**Pull request by elee1766**
address #2783

still wip.

features:
[x] single pass
[x] single write to stdout
[x] errors to stderr
[x] parses flags in  %[flags][width][.precision][length]specifier format, with support for *
[x] \" \\ \a \b \c \e \f \n \r \t \v \NNN \0NNN \xHH \uHHHH \uHHHHHHHH
[x] %b
[x] %s
[x] %q
[-] %d partially implemented, missing width, precision, length, padding, etc. can parse hex strings. 
[ ] {iouxXfeEgGc}

### CODE REVIEW

**Comment by elee1766** [comment]
@rminnich hi

i had some thoughts while implementation, thought you would be a good person to give opinions, sorry if you are not the right person to ping for this 

first of all,  each implementation seems to deals with printf differently. they not only have different features for arguments, but also support different escape codes, etc. 

for instance, 
`printf "%d" "123+123"` is valid in only `zsh` , because `zsh` supports math expressions
`printf \u1230` is not supported by debian `sh` and simply returns `\u1230` 
`printf \u123` errors in coreutils, pads to `\u0123` in bash & zsh
`printf --help` simply returns `--help` in `zsh`, while returning help menu for bash & coreutils
`printf %q` is different for every shell, because it formats in a way that is friendly to the shell that it is in
the list continues on and on

as a result, i think the best way to do printf would be...

1. move cmd/printf to pkg/printf
2. create cmd/printf which mirrors the behavior of coreutils printf
3. create printf builtin in `gosh`, which has format code %q/%Q for gosh semantics, and can add any shell specific integrations

i'm not too familiar with gosh,  if it is fully posix compliant, then we can probably share %q formatters, and just need to make sure that input args get properly encoded/passed in a friendly manner

File: cmds/core/printf/printf.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+import (
+	"bytes"
+	"io"
+	"os"
+)
+
+type printf struct {
+	Args   []string
+	Stdout io.Writer
```
Comment (rminnich) [inline cmds/core/printf/printf.go:68]: since you have a new, is there a reason to export these? Does some package require it? I don't know.

File: cmds/core/printf/printf.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+package main
+
+import (
+	"bytes"
+	"io"
+	"os"
+)
+
+type printf struct {
```
Comment (rminnich) [inline cmds/core/printf/printf.go:66]: another thing I don't know. Does this type implement some interface? If it does, even though lint does not require it, a comment might be helpful for those who come later.

File: cmds/core/printf/printf.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	Args   []string
+	Stdout io.Writer
+	Stderr io.Writer
+
+	format    string
+	arguments []string
+}
+
+func NewPrinter(stdout, stderr io.Writer, args []string) *printf {
```
Comment (rminnich) [inline cmds/core/printf/printf.go:75]: // NewPrinter returns a printf

or some kind of comment.

File: cmds/core/printf/printf.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		Args:   args,
+	}
+}
+
+func (c *printf) exec(w *bytes.Buffer) (err error) {
+	return interpret(w, c.format, c.arguments, false, true)
+}
+
+func (c *printf) run() {
+	if len(c.Args) < 1 {
```
Comment (rminnich) [inline cmds/core/printf/printf.go:88]: The command way to handle this would be, in the New function, to have an Arg0 and Args.
Or, further, set  c.format in the New. By the time printf is called, this kind of arg checking ought to have been done, and New is a good place for it.

File: cmds/core/printf/printf.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		c.Stderr.Write([]byte("printf: not enough arguments\n"))
+		return
+	}
+	c.format = c.Args[0]
```
Comment (rminnich) [inline cmds/core/printf/printf.go:92]: This could be done in the New

File: cmds/core/printf/printf.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	}
+}
+
+func (c *printf) exec(w *bytes.Buffer) (err error) {
+	return interpret(w, c.format, c.arguments, false, true)
+}
+
+func (c *printf) run() {
+	if len(c.Args) < 1 {
+		c.Stderr.Write([]byte("printf: not enough arguments\n"))
```
Comment (rminnich) [inline cmds/core/printf/printf.go:89]: Would be nice if this could return an error, not do this write.

File: cmds/core/printf/printf.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		return
+	}
+	c.format = c.Args[0]
+	if len(c.Args) > 1 {
```
Comment (rminnich) [inline cmds/core/printf/printf.go:93]: This could be set in New.

File: cmds/core/printf/printf.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		c.arguments = c.Args[1:]
+	}
+	w := new(bytes.Buffer)
+	err := c.exec(w)
+	if err != nil {
+		c.Stderr.Write([]byte("printf: " + err.Error() + "\n"))
```
Comment (rminnich) [inline cmds/core/printf/printf.go:99]: This can return an error, it makes tests easier. Also, you can use fmt.Errorf, with a wrapped error type (%w) which makes your tests simpler.

File: cmds/core/printf/printf_test.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	args := func(args ...string) []string {
+		return args
+	}
+
+	type testCase struct {
+		a   []string
+		e   string
+		err string
```
Comment (rminnich) [inline cmds/core/printf/printf_test.go:24]: if you have run return an error, you can use wrapper errors, and the err here can be an error type. That is preferred for tests now.

Comment (rminnich) [reply cmds/core/printf/printf_test.go:24]: you can use errors.Is.

File: cmds/core/printf/utils.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+func interpret(
```
Comment (rminnich) [inline cmds/core/printf/utils.go:10]: in general, in Go, this kind of arg formatting is not really idiomatic. Please fit this declaration on one line.

File: cmds/core/printf/utils.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+				continue
+			case 'G':
+				continue
+			case 'c':
+				continue
+			case 's':
+				fmt.Fprintf(o, "%s", arg)
+				continue
+			default:
+				return fmt.Errorf("%s: %s", "%"+string(n), "invalid directive")
```
Comment (rminnich) [inline cmds/core/printf/utils.go:87]: if you can make an error: 
var errIInvalidDirective = errors.New("invalid directive")
and then in the fmt.Errorf, wrap it with %w, then your tests are easier to write.

File: cmds/core/printf/utils.go (hunk 10 above)
Comment (rminnich) [inline cmds/core/printf/utils.go:10]: I'm a bit surprised to see you had to write your own interpreter, is the one in fmt not usable for some reason?

File: cmds/core/printf/utils.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+			o.WriteRune(n)
+		} else {
+			// otherwise just write the rune
+			o.WriteRune(c)
+		}
+	}
+	return nil
+}
+
+func readOctal(fr *strings.Reader, o *bytes.Buffer) {
```
Comment (rminnich) [inline cmds/core/printf/utils.go:163]: I am fairly certain you can use strconv package for this, just set the base.

File: cmds/core/printf/utils.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+			continue
+		}
+		// not an decimal, so unwind the rune and also break, printing the octal if it exists
+		fr.UnreadRune()
+		break
+	}
+	if len(hexcode) == 0 {
+		o.WriteRune('\\')
+	} else {
+		i, err := strconv.ParseInt(hexcode, 16, 32)
```
Comment (rminnich) [inline cmds/core/printf/utils.go:206]: I'm not sure I see the need for this. Am I missing something (likely :-)?

File: cmds/core/printf/printf.go (hunk 1 above)
Comment (elee1766) [reply cmds/core/printf/printf.go:68]: nope it was just to make clear for me which were constructor args and which weren't while I was writing

File: cmds/core/printf/utils.go (hunk 10 above)
Comment (elee1766) [reply cmds/core/printf/utils.go:10]: the fmt in golang is semantically different from every implementation that I compared with (coreutils, bash, zsh, sh) 

I'm pretty sure this will be inevitable, especially if we need to implement format code %q

File: cmds/core/printf/utils.go (hunk 12 above)
Comment (elee1766) [reply cmds/core/printf/utils.go:163]: I do within readoctal, but we need to be able to stop reading the octal after either N characters OR upon the first non decimal character, which I can't do with just strconv

File: cmds/core/printf/utils.go (hunk 13 above)
Comment (elee1766) [reply cmds/core/printf/utils.go:206]: do you mean the error check? I don't think the error check is needed here, I'm not sure if it's better to add a comment explaining why, or just handle it

Comment (elee1766) [reply cmds/core/printf/utils.go]: oh do you mean the check if hexcode is 0? i think that bit is wrong actually :-)

File: cmds/core/printf/printf.go (hunk 2 above)
Comment (elee1766) [reply cmds/core/printf/printf.go]: no, but maybe it should. 

i thought briefly about making printf an io.Reader instead of accepting an io.Writer as input, but that felt wrong.

**Comment by elee1766** [comment]
@rminnich i've changed the error handling, moved printf to another package, and added a few comments. 

as to reasons why i had to start on a custom interpreter instead of go fmt, here are a few - 
1. some format codes, e.g. %q, %b have completely different meanings
2. some format codes, e.g. %x, %X, %p have slightly different rules 
3. go is missing format code %i
4. go is missing escape sequence \e

along with that, parsing arguments is completely different. for instance, `printf '%d' 0x1234' interprets the hex and will output `4660` and `printf '%s' 0x1234` will keep 0x1234 as a string, which go's fmt can't do. 

basically, there will actually need to be helper functions which parse the parameters into their correct form based off the format code they match with

after that though, we can reuse some logic and use fprintf with the go directive, as i am with %s

**Comment by elee1766** [comment]
ok, i've created a framework to create the rest of the format codes.

**Comment by rminnich** [comment]
I think this person is not coming back.
